### PR TITLE
feat(#2267): create new checkbox list component

### PR DESCRIPTION
**Pull request by syedszeeshan**
# Before (the change)

# After (the change)

The video shows the new Checkbox list component with various examples. e.g. 

- disabled state
- error state and programatically toggle b/w error and normal state
- select all with indeterminate
- required with error and helper text
- (Angular specific examples) reactive and template driven forms with clear/reset functioanlity
- Public form intergration
- Public form summary showing selected ~~values~~ text
- Showing one item per line when multiple values selected

https://github.com/user-attachments/assets/0ee5cd2f-861e-417a-98c3-a94ab8af3980

## Make sure that you've checked the boxes below before you submit the PR

- [x] I have read and followed the [setup steps](https://goa-dio.atlassian.net/wiki/spaces/DS/pages/3086385489/Contribution+-+web+content#%5BinlineExtension%5DCode-contributions)
- [x] I have created necessary unit tests
- [x] I have tested the functionality in both React and Angular.

## Steps needed to test

**Playground Code**
**Angular**
[checkbox-list.component.html..txt](https://github.com/user-attachments/files/22244283/checkbox-list.component.html.txt)
[checkbox-list.component.html.ts.txt](https://github.com/user-attachments/files/22244284/checkbox-list.component.html.ts.txt)

**React**
[app.tsx.txt](https://github.com/user-attachments/files/22244512/app.tsx.txt)
[updated-app.tsx.txt](https://github.com/user-attachments/files/22260853/updated-app.tsx.txt)

### CODE REVIEW

File: libs/common/src/lib/common.ts [hunk 1]
```diff
@@ -7,6 +7,14 @@ export type GoabRadioGroupOnChangeDetail = {
   value: string;
 };
 
+// Checkbox List
+export type GoabCheckboxListOrientation = "vertical" | "horizontal";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/common/src/lib/common.ts]: I have removed orientation property.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+      })
+      .filter((el): el is HTMLElement => !!el);
+  }
+
+  /**
+   * Determines a child's identifier by trying expected attributes.
+   * Priority: name -> value
+   */
+  function getCheckboxIdentifier(el: HTMLElement): string {
+    //return (el as any).name || (el as any).value || "";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: comment removed.

File: libs/react-components/src/lib/checkbox-list/checkbox-list.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
+import {
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/react-components/src/lib/checkbox-list/checkbox-list.tsx:1]: PR has been updated.

File: libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    },
+  ],
+  imports: [NgTemplateOutlet, NgIf],
+})
+export class GoabCheckboxList extends GoabControlValueAccessor {
+  @Input() name?: string;
+  @Input() ariaLabel?: string;
+  @Input() description!: string | TemplateRef<unknown>;
+  @Input() maxWidth?: string;
+  @Input() childml?: Spacing;
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts]: PR has been updated.

File: libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
+import { GoabCheckboxListOnChangeDetail, Spacing } from "@abgov/ui-components-common";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts:1]: yes, working on it now.

[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts:1]: Browser test has been added now.

File: libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    },
+  ],
+  imports: [NgTemplateOutlet, NgIf],
+})
+export class GoabCheckboxList extends GoabControlValueAccessor {
+  @Input() name?: string;
+  @Input() ariaLabel?: string;
+  @Input() description!: string | TemplateRef<unknown>;
+  @Input() maxWidth?: string;
+  @Input() childMl?: Spacing;
```
Comment (chrisolsen) [inline libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts:55]: I thought this was going to be removed some the "Select/Deselect All" was going to be removed

File: libs/angular-components/src/lib/components/checkbox/checkbox.ts [hunk 7]
```diff
@@ -56,6 +57,7 @@ import { GoabControlValueAccessor } from "../base.component";
 export class GoabCheckbox extends GoabControlValueAccessor {
   @Input() name?: string;
   @Input({ transform: booleanAttribute }) checked?: boolean;
+  @Input({ transform: booleanAttribute }) indeterminate?: boolean;
```
Comment (chrisolsen) [inline libs/angular-components/src/lib/components/checkbox/checkbox.ts:60]: Since indeterminate is not being used, this property should not be exposed.

File: libs/web-components/src/components/form/FormSummary.svelte [hunk 8]
```diff
@@ -104,18 +104,53 @@
...
+      const [year, month, day] = stringValue.split("-");
       return `${months[parseInt(month) - 1]} ${parseInt(day)}, ${year}`;
     }
-    return value;
+    return stringValue;
+  }
+
+  function formatCsvAsMultiline(
+    value: string | number | Date,
+    valueLabel?: string,
```
Comment (chrisolsen) [inline libs/web-components/src/components/form/FormSummary.svelte:127]: We also talked about this. I mentioned it has to be in a JSON format to prevent unwanted splitting of strings that contain a `,`

**Comment by syedszeeshan** [comment]
> The commit text also needs to be changed to follow the recomendations

commit message updated to sentence style.

File: libs/web-components/src/components/form/FormSummary.svelte (hunk 8 above)
Comment (syedszeeshan) [reply libs/web-components/src/components/form/FormSummary.svelte]: formatValue function updated to handle string[], the formatCsvAsMultiline function remvoved.

File: libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts (hunk 6 above)
Comment (syedszeeshan) [reply libs/angular-components/src/lib/components/checkbox-list/checkbox-list.ts]: childMl prop ha been removed from .svelte web comp, angular and react wrappers

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+  }
+
+  // Update the selected values array when an individual child changes.
+  // Consolidate child checkbox changes into the group's selected array and emit a single
+  // _change with both the raw values and a user-friendly optionLabel string.
+  function handleChildCheckboxChange(detail: any) {
+    const checkboxName = detail.name;
+
+    const isChecked =
+      typeof detail.checked === "boolean" ? detail.checked : !!detail.value;
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: As discussed, using detail.checked now.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+
+  type ChildRecord = { el: HTMLElement; name: string; label?: string }; // label is captured to build optionLabel for the form summary.
+
+  // Private state
+  let _rootEl: HTMLElement;
+  let _slotEl: HTMLElement;
+  let _error: boolean;
+  let _prevError: boolean;
+  let _childRecords: ChildRecord[] = [];
+  let _allCheckboxValues: string[] = [];
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: PR has been updated

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+  export let description: string = "";
+  export let maxwidth: string = "none";
+
+  export let mt: Spacing = null;
+  export let mr: Spacing = null;
+  export let mb: Spacing = "m";
+  export let ml: Spacing = null;
+
+  // Optional left margin to apply to each child checkbox
+  export let childml: Spacing = null;
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: PR has been updated.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+  export let disabled: string = "false";
+  export let error: string = "false";
+  export let testid: string = "";
+  export let arialabel: string = "";
+  export let description: string = "";
+  export let maxwidth: string = "none";
+
+  export let mt: Spacing = null;
+  export let mr: Spacing = null;
+  export let mb: Spacing = "m";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: PR has been updated

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+  function handleChildCheckboxChange(detail: any) {
+    const checkboxName = detail.name;
+
+    const isChecked =
+      typeof detail.checked === "boolean" ? detail.checked : !!detail.value;
+    let newSelectedValues = [...value];
+
+    if (isChecked) {
+      if (!newSelectedValues.includes(checkboxName))
+        newSelectedValues.push(checkboxName);
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: PR has been updated.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+    await tick();
+    addRelayListener();
+    addSlotEventListeners();
+    sendMountedMessage();
+
+    // Initialize after a tick to ensure DOM is ready
+    _isInitialized = true;
+    // Only sync if we actually have child records, otherwise wait for them to mount
+    if (_childRecords.length > 0) {
+    }
```
Comment (chrisolsen) [inline libs/web-components/src/components/checkbox-list/CheckboxList.svelte:64]: Nothing in this `if` block

File: libs/react-components/src/lib/checkbox-list/checkbox-list.tsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/react-components/src/lib/checkbox-list/checkbox-list.tsx:1]: @vanessatran-ddi 
PR has been updated, and is ready for testing.

File: libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+      });
+
+      // Child should be updated to checked state via relay message
+      expect(checkbox.getAttribute("name")).toBe("option1");
+    });
+  });
+
+  describe("Integration with slotted content", () => {
+    it("should handle slotted checkboxes", async () => {
+      const Component = {
```
Comment (chrisolsen) [inline libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts:202]: This test initially got me really excited, in that I thought that more easily create complete parent/child components for unit testing, but looking further the `slots` prop is never used and the component `Component` and `props` are just directly referenced. So my comment at the top of this file still stands, in that we need to use the browser tests.

File: libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts [hunk 17]
```diff
@@ -0,0 +1,242 @@
+import { cleanup, render, waitFor } from "@testing-library/svelte";
```
Comment (chrisolsen) [inline libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts:1]: This component needs browser tests since none of these tests have child Checkbox elements.

File: libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+
+  describe("Child checkbox management", () => {
+    it("should update child checkbox states based on value", async () => {
+      const { container } = render(CheckboxList, {
+        ...defaultProps,
+        value: ["option1"],
+      });
+
+      // Add child checkbox
+      const checkboxContainer = container.querySelector(".checkbox-container");
```
Comment (chrisolsen) [inline libs/web-components/src/components/checkbox-list/CheckboxList.spec.ts:172]: A lot of these tests are very fragile in that they are obtaining elements vis class selectors.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+      if (detail && detail.value !== undefined) {
+        handleChildCheckboxChange(detail);
+      }
+    });
+  }
+
+  // Update the selected values array when an individual child changes.
+  // Consolidate child checkbox changes into the group's selected array and emit a single
+  // _change with both the raw values and a user-friendly optionLabel string.
+  function handleChildCheckboxChange(detail: any) {
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte:294]: As suggested, now using Set, PR has been updated.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+      _rootEl?.dispatchEvent(
+        new CustomEvent("error::change", {
+          detail: { isError: currentError },
+          bubbles: true,
+          composed: true,
+        }),
+      );
+      _error = currentError;
+      updateChildCheckboxesError();
+    }
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: code block removed, PR has been updated.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+
+    _rootEl?.dispatchEvent(
+      new CustomEvent("_change", {
+        detail: {
+          name,
+          value: newSelectedValues,
+          optionLabel: buildOptionLabel(newSelectedValues),
+        },
+        bubbles: true,
+        composed: true,
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte:315]: I have renamed `buildOptionLabel `to a more meaningful name "`getSelectedLabels`"

However, if I change "optionLabel" (see strongly-typed example below ) then this will have to be changed in FieldSet also and RadioGroup, b/c they are using and expecting optionLabel


```
interface CheckboxListChangeDetail {
  name: string;
  value: string[];
  labels: string[]; // renamed from optionLabel
}

_rootEl?.dispatchEvent(
  new CustomEvent<CheckboxListChangeDetail>("_change", {
    detail: {
      name,
      value: newSelectedValues,
      labels: buildOptionLabel(newSelectedValues), // renamed property
    },
    bubbles: true,
    composed: true,
  }),
);
```

Do you still want me to change optionLabel to labels?

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 22]
```diff
@@ -0,0 +1,415 @@
...
+      element.setAttribute("checked", shouldBeChecked ? "true" : "false");
+      if (isDisabled) element.setAttribute("disabled", "true");
+      else element.removeAttribute("disabled");
+    });
+  }
+
+  /** Sync a single child checkbox with the parent state and disabled flag. */
+  function updateChildCheckboxState(childEl: HTMLElement, childName?: string) {
+    const name =
+      childName || _childRecords.find((r) => r.el === childEl)?.name || "";
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: I have removed this, PR has been updated.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 23]
```diff
@@ -0,0 +1,415 @@
...
+
+    if (name) {
+      const shouldBeChecked = value.includes(name);
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte]: 1. This has been updated to below
`const shouldBeChecked = _selectedValues.has(checkboxName);`

2. Example"

```
<goa-checkbox-list name="contact_preferences" value='["email", "phone"]'>
  <goa-checkbox name="email" text="Email"></goa-checkbox>     ← should be checked
  <goa-checkbox name="phone" text="Phone"></goa-checkbox>     ← should be checked  
  <goa-checkbox name="sms" text="SMS"></goa-checkbox>         ← should NOT be checked
</goa-checkbox-list>
```

For checkbox [name="email"]: ["email", "phone"].includes("email") → true ✅
For checkbox [name="phone"]: ["email", "phone"].includes("phone") → true ✅
For checkbox [name="sms"]: ["email", "phone"].includes("sms") → false ❌

I will connect with you in a screenshare, and check if there is a better way.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+      relay<FieldsetSetValueRelayDetail>(childEl, FieldsetSetValueMsg, {
+        name,
+        value: shouldBeChecked ? "checked" : "",
+      });
+    }
+
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte:359]: "checked" made more sense to me but 
I can pass string "true" or string "1" or may be the value itself.... as long as it is any non-empty string, will convert to truthy value

or we can also try shouldBeChecked (boolean)

I will connect with you in a screenshare.

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte (hunk 21 above)
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte:329]: As discussed, 
-the optionLabel has been changed to label
-labels has been added to pass string array

File: libs/web-components/src/components/checkbox-list/CheckboxList.svelte (hunk 20 above)
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/web-components/src/components/checkbox-list/CheckboxList.svelte:101]: As discussed, keeping this code for now.

File: libs/angular-components/src/lib/components/checkbox/checkbox.ts [hunk 25]
```diff
@@ -18,11 +18,12 @@ import { GoabControlValueAccessor } from "../base.component";
   selector: "goab-checkbox",
```
Comment (ArakTaiRoth) [inline libs/angular-components/src/lib/components/checkbox/checkbox.ts:18]: I'm not sure where this error is coming from, it might not have to do with this file. I'm just logging it here because it's an Angular specific issue.

If you setup a normal Checkbox using an Angular FormControl and FormGroup (as shown in our documentation), the checkbox returns boolean values `true` and `false`, instead of returning "checked" or null like it does in Alpha. In addition, the boolean values are wrong, it returns `true` when unchecked and `false` when checked.

Comment (syedszeeshan) [reply libs/angular-components/src/lib/components/checkbox/checkbox.ts]: PR has been updated, reverted couple of changes from the ~~`checkbox-list.ts`~~  `checkbox.ts` angular wrapper as they were not needed anymore. The checkbox.ts is pretty much back to how it was before, except the `indeterminate `prop.

File: libs/react-components/src/lib/checkbox-list/checkbox-list.tsx [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
```
Comment (ArakTaiRoth) [inline libs/react-components/src/lib/checkbox-list/checkbox-list.tsx:1]: This is a React only issue so I've tagged it in this file. The values in React (`GoabCheckboxListOnChangeDetails.value`) are returned in one non-broken string. This would be impossible to parse. In Angular it's returned comma-delimited, which is perfectly fine and even expected for parsing.
Angular:
image

React:
image

Comment (syedszeeshan) [reply libs/react-components/src/lib/checkbox-list/checkbox-list.tsx:1]: https://github.com/user-attachments/assets/acd98137-9fe6-4e98-a4a7-c9c206a6eca0

Updated the react aplayground also and attached.

As discussed, still returning an array. React (by default) is not adding a separator for browser display, but Json.stringify or .join can be used to display in the desired format.

File: libs/react-components/src/lib/checkbox-list/checkbox-list.tsx [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+  name: string;
+  value?: string[];
+  disabled?: string;
+  error?: string;
+  testid?: string;
+  maxwidth?: string;
+}
+
+export interface GoabCheckboxListProps extends Margins {
+  name: string;
```
[the remark this answers is not in the document]
Comment (syedszeeshan) [reply libs/react-components/src/lib/checkbox-list/checkbox-list.tsx:27]: PR has been updated.

File: libs/web-components/src/components/checkbox/Checkbox.svelte [hunk 28]
```diff
@@ -149,12 +156,15 @@
...
-      { bubbles: true, timeout: 10 },
-    );
+    const checkboxEl = (_rootEl.getRootNode() as ShadowRoot)?.host as HTMLElement;
+    const fromCheckboxList = checkboxEl?.closest("goa-checkbox-list") !== null;
+
+      relay<FormFieldMountRelayDetail>(
+        _rootEl,
+        FormFieldMountMsg,
+        { name, el: _rootEl },
+        { bubbles: !fromCheckboxList, timeout: 10 },
```
[the remark this answers is not in the document]
Comment (chrisolsen) [reply libs/web-components/src/components/checkbox/Checkbox.svelte:165]: This is breaking the isolation design pattern. This component should not be touched to work with the CheckboxList (excluding the indeterminate prop). Instead the checkbox list should handle the `mount` messages from the child checkboxes and stop them from bubbling up any further.

**Review by chrisolsen — APPROVED** [review]
